cli/data.go

```diff
@@ -0,0 +1,194 @@
+package cli
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	datapb "go.viam.com/api/app/data/v1"
+	rdkutils "go.viam.com/rdk/utils"
+	"google.golang.org/protobuf/encoding/protojson"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+const (
+	dataDir     = "data"
+	metadataDir = "metadata"
+)
+
+// BinaryData writes the requested data to the passed directory.
+func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
+	fmt.Println(filter.OrgIds)
+
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	// TODO: Should we use more limited perms?
+	// TODO: Probably shouldn't re-download files we already have? Maybe we actually should have an exlclude_ids field
+	if err := os.MkdirAll(filepath.Join(dst, dataDir), os.ModePerm); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Join(dst, metadataDir), os.ModePerm); err != nil {
+		return err
+	}
+
+	skip := int64(0)
+	for {
+		resp, err := c.dataClient.BinaryDataByFilter(context.Background(), &datapb.BinaryDataByFilterRequest{
+			DataRequest: &datapb.DataRequest{
+				Filter: filter,
+				Skip:   skip,
+				Limit:  1,
+			},
+			IncludeBinary: true,
+			CountOnly:     false,
+		})
+		// TODO: Make sure EOF is properly interpreted. Iirc rpc errors aren't properly parsed by errors.Is.
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		mds := resp.GetMetadata()
+		if len(mds) != 1 {
+			return errors.Errorf("expected a single response, received %d", len(mds))
+		}
+		data := resp.GetData()
+		if len(data) != 1 {
+			return errors.Errorf("expected a single response, received %d", len(data))
+		}
+
+		md := mds[0]
+		datum := data[0]
+		mdJsonBytes, err := protojson.Marshal(md)
+		if err != nil {
+			return err
+		}
+		jsonFile, err := os.Create(filepath.Join(dst, "metadata", datum.GetId()+".json"))
```

Comment:
Would ideally keep this consistent between tabular and binary. In binary data this is metadata/{datum_id}.json, whereas in tabular data this is metadata/{metadata_index}.json

This current loop will create a lot of repeated metadata messages, so I wonder if it's worth storing in a set and adding a metadata index to make consistent with tabular data.

Reply:
As I work on this, I'm actually starting to think maybe we shouldn't be doing the same metadata_index stuff we do with tabular. For tabular data, metadata is similar in size or can even exceed the size of the actual data, so by repeating it we'd be being super inefficient. For binary data, that's not the case, and I'm finding it fairly awkward to merge the shared and datum level metadata

Reply:
We would want the user to easily be able to decode what's in `/data` and `/metadata`, so I assume you mean merging the two metadata messages within the actual API.

Thinking through what you said, I'm leaning towards your suggestion. Tabular is 1 metadata -> 1 group of sensor messages which can be combined into a CSV, whereas binary data is 1 metadata -> 1 group of images which each has additional datum-level information that can't be combined.

Reply:
Agree to merge the two in JSON format here, but then would think about changing the API to merge within the proto.